### PR TITLE
Code review for free()

### PR DESCRIPTION
**Pull request by georgit**
Still figuring out pull requests. But at least I didn't make a georgit branch this time. So there's that.

### CODE REVIEW

**Comment by gparmer** [comment]
You'll need to start making more, finer-grained commits.  Your commit messages and names don't really give much of an indication which feature your touching, or what the commit is doing.

File: src/components/Makefile.comp [hunk 1]
```diff
@@ -35,8 +35,8 @@ INC_PATH=-I./ -I$(CDIR)/include/ -I$(CDIR)/interface/
 # probably ballooning its size.  Fix this need for no merge constants
 # in cos_loader.
 SHARED_FLAGS=-fno-merge-constants -nostdinc -nostdlib
-#OPT= -ggdb3
-OPT= -O3
+OPT= -ggdb3
```
Comment (gparmer) [inline src/components/Makefile.comp]: Please avoid tangential changes.

File: src/components/implementation/tests/unit_cbufmalloc/malloc.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#define NUM 1024
+char *mptr[NUM];
+
+int
+test_malloc(void)
+{
+	int i, j;
+
+	for(i=0; i<NUM; i++) {
+        printc("Call to malloc:\n");
```
Comment (gparmer) [inline src/components/implementation/tests/unit_cbufmalloc/malloc.c]: Why is this inconsistently spaced?  Looks like this line uses spaces, and the rest use tabs.

File: src/components/implementation/tests/unit_cbufmalloc/malloc.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		mptr[i] = (char *)malloc(i+1);
+		assert(mptr[i]);
+		for(j=0; j<i; j++) mptr[i][j] = '$';
+		//free(mptr[i]);
```
Comment (gparmer) [inline src/components/implementation/tests/unit_cbufmalloc/malloc.c]: Aren't you testing `free`?  Shouldn't this be uncommented?

File: src/components/implementation/tests/unit_cbufmalloc/posix_test.c [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        for (j = 0; j < i; j++)
+        {
+            mptr[i][j] = j;
+        }
+        
+        for (j = 0; j < i; j++)
+        {
+            printc("%d,", mptr[i][j]);
+        }
+        printc("\n");
```
Comment (gparmer) [inline src/components/implementation/tests/unit_cbufmalloc/posix_test.c]: Your spacing is completely jacked throughout this function.  I can't read this on github.

File: src/components/implementation/tests/unit_cbufmalloc/posix_test.c [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		free(mptr[i]);
+	}
+
+	return !!err;
```
Comment (gparmer) [inline src/components/implementation/tests/unit_cbufmalloc/posix_test.c]: How is this not returning an error?  `return` statement in a function that returns `void`.

File: src/components/implementation/tests/unit_posix/malloc.c [hunk 6]
```diff
@@ -16,8 +16,9 @@ test_malloc(void)
 		mptr[i] = (char *)malloc(i+1);
 		assert(mptr[i]);
 		for(j=0; j<i; j++) mptr[i][j] = '$';
-		free(mptr[i]);
```
Comment (gparmer) [inline src/components/implementation/tests/unit_posix/malloc.c]: I'm not convinced that you want to be weakening the testing regime here.

File: src/components/implementation/tests/unit_posix/posix_test.c [hunk 7]
```diff
@@ -34,27 +34,27 @@ cos_init(void *args)
...
 	RUN_TEST(malloc);
-	RUN_TEST(memstream);
-	RUN_TEST(qsort);
-	RUN_TEST(sscanf);
-	RUN_TEST(snprintf);
-	RUN_TEST(string);
-	RUN_TEST(strtod);
-	RUN_TEST(strtol);
-	RUN_TEST(wcstol);
+	//RUN_TEST(memstream);
```
Comment (gparmer) [inline src/components/implementation/tests/unit_posix/posix_test.c]: Again, don't weaken our testing.

File: src/components/interface/cbuf_mgr/cbuf_c.c [hunk 8]
```diff
@@ -136,7 +136,7 @@ __cbufp_alloc_slow(int cbid, int size, int *len, int *error)
 	/* Nothing collected...allocate a new cbuf! */
 	if (amnt == 0) {
 		cbid = cbuf_create(cos_spd_id(), size, cbid*-1);
-		if (cbid == 0) assert(0);
+		if (cbid == 0) {assert(0)};
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cbuf_c.c]: If you want to use `{}`, then spread it out to three lines.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+		prev = __small_mem[idx];
+		ptr->next=prev;
+	} while (unlikely(cos_cmpxchg(&__small_mem[idx], (int)prev, (int)ptr) != (int)ptr));
+	alloc_stats_report(DBG_FREE, idx);
+}
+
+static void* do_cbuf_alloc(size_t size, cbuf_t *cbid)
+{
+    if (!size)
+    {
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: `{` only on a separate line for functions.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+        return NULL;
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: I would just put these all on the same line `if (!size) return NULL` as it is an error case, and you want to de-emphasize it in the code.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+	alloc_stats_report(DBG_ALLOC, idx);
+
+	return ptr;
+}
+
+/* -- PUBLIC FUNCTIONS ---------------------------------------------------- */
+
+static void do_cbuf_free(cbuf_t cbid)
+{
+    printc("        calling cbuf free\n");
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: Remove debugging prints when you do pulls/commits "for keeps".

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+    cbuf_free(cbid);
+}
+
+static void _alloc_libc_free(void *ptr) 
+{
+    printc("    calling free\n");
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: ditto....

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 13]
```diff
@@ -0,0 +1,536 @@
...
+    if (ptr)
+    {
+        printc("    found a pointer\n");
+        __alloc_t *pointer = BLOCK_START(ptr);
+        size_t size = pointer->size;
+
+        printc("    size is %zu\n", size);
+        if (size)
+        {
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: Same spacing problem.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 14]
```diff
@@ -0,0 +1,536 @@
...
+{
+    printc("        calling cbuf free\n");
+    cbuf_free(cbid);
+}
+
+static void _alloc_libc_free(void *ptr) 
+{
+    printc("    calling free\n");
+    if (ptr)
+    {
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: spacing problem.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 15]
```diff
@@ -0,0 +1,536 @@
...
+void *alloc_page(void)
+{
+	struct free_page *fp;
+	void *a;
+
+	fp = page_list.next;
+	if (NULL == fp) {
+        cbuf_t cbid;
+		a = do_cbuf_alloc(PAGE_SIZE, &cbid);
+        // what if this fails?
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: `return NULL`  In systems code, you always have to check the return value of memory allocation calls, and deal with the contingencies.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 16]
```diff
@@ -0,0 +1,536 @@
...
+      ptr=_alloc_libc_malloc(size);
+    }
+  }
+  return ptr;
+}
+void* realloc(void* ptr, size_t size) __attribute__((weak,alias("__libc_realloc")));
+
+#endif
+
+/********************* testing code on unix ***********************/
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: You don't really support any of this Linux, user-level testing, so you might as well remove it.

File: src/components/interface/cbuf_mgr/cos_alloc.c [hunk 17]
```diff
@@ -0,0 +1,536 @@
...
+	}
+
+	free((void*)*ptr);
+	*ptr = 0;
+	printf("-");
+
+	return;
+}
+
+int main(void)
```
Comment (gparmer) [inline src/components/interface/cbuf_mgr/cos_alloc.c]: You might consider testing malloc with something that looks like this, though.  Seems more complete than your tests.  `rand()` will need to be emulated with `rdtscll`.

**Comment by gparmer** [comment]
Good first push.  There isn't much code here, but you spent most of your time learning the code-base, and setting up your environment.  Now that you're past those humps, you'll find your productivity will rise.

**Comment by georgit** [comment]
Very strange about spacing. I must not have something spaced out right.
Also, sorry about the commit messages. I know there is a `git squash`
function or something similar that puts all of your commits together. I was
going to do that and then write you a much better commit message for
everything in that branch but then I sort of forgot (still a good thing
given Unity crashed shortly after). The malloc.c test was supposed to be
deleted from cbufMalloc testing and the code moved to posix_test.c (which I
should probably rename). I'm surprised you even see that file...

Will fix and then resubmit.

-Teo

**Pull request closed without merging**